Robot/neopixel-leds/src/test/java/org/usfirst/frc/team4999/tools/BufferDisplay.java

```diff
@@ -0,0 +1,175 @@
+     * The BufferDisplay is a Display which will interpret received packets and draw the animations to a Color[] buffer.
+     * The BufferDisplay implements the Observable design pattern. 
+     * Upon receipt of a "show" packet, the BufferDisplay will notify all observers with a copy of the rendered Color[] buffer.
+     * It is intended to keep all the code for parsing and interpreting packets in one file.
+     * @param numPix
+     */
+    public BufferDisplay(int numPix) {
+        buffer = new Color[numPix];
+        for(int i = 0; i < buffer.length; i++) {
+            buffer[i] = Color.BLACK;
+        }
+    }
+
+    public void addBufferListener(BufferUpdateListener listener) {
+        listeners.add(listener);
+    }
+
+    public void removeBufferListener(BufferUpdateListener listener) {
+        listeners.remove(listener);
+    }
+
+    private void notifyBufferListeners() {
+        for(BufferUpdateListener listener : listeners) {
+            listener.onBufferUpdated(Arrays.copyOf(buffer, buffer.length));
+        }
+    }
+    
+    private int unsignedByteValue(byte b) {
+        // See https://stackoverflow.com/questions/11380062/what-does-value-0xff-do-in-java
+        // Basically, when the byte is cast to an int, it is sign-extended.
+        // By masking only the last byte, we remove the sign-extended bits.
+        return ((int) b) & 0xff;
+    }
+
+    private void interpretPacket(Packet packet) {
+        /*
+        This interpretation code should be identical to that of the arduino
+
+        Packet format:
+        - B0 : payload size
+        - B1-15 : payload  (Max payload is 15 bytes)
+
+        Command payload format:
+        - B0 : command id
+        - B1-14 : command data (varies by command)
+
+        Commands:
+        - 0x01 : <empty> : show current image
+        - 0x02 : pixel address ; RGB : set single pixel at address with value RGB
+        - 0x03 : start address ; RGB ; length : set "length" pixels starting at "start" with value RGB
+        - 0x04 : start address ; RGB ; length ; stride : set "length" pixels starting at "start" with value RGB.  Repeat every "stride" pixels.
+        - 0x05 : start address ; RGB ; length ; stride ; max length : same as 0x04, but will stop after max length pixels
+
+        Error rejection/correction:
+        - Invalid packets will be rejected and packet parsing will begin again with next byte in stream.
+
+        Display protocol:
+        - At beginning of frame, send >= 16 0xff bytes to force resync
+        - Send paint commands to fill image buffer.  Unpainted pixels will retain prior contents.
+        - Send "show current image" command
+
+        */
+        byte[] rawPacketBytes = packet.getData();
+        int payloadLen = unsignedByteValue(rawPacketBytes[0]);
+        if(rawPacketBytes.length - 1 != payloadLen) {
+            System.err.format("Invalid packet: stated length %d does not match actual length %d\n", payloadLen, rawPacketBytes.length - 1);
```

Comment:
If this code did receive a resync packet, it would spam the log. Maybe I should check for that?
However, this condition should never be reached since resync packets are sent by the NeoPixels display, not by the animator, so this code should never receive a Resync packet.

Reply:
I do think it should handle resync gracefully since it's specified as part of the protocol.
The bigger question is if/how to test it since this environment would never get out of sync.